### PR TITLE
net/buf: Add the possibility to manipulate externally allocated buf/data

### PR DESCRIPTION
**Pull request by tbursztyka**
Note that name sucks a bit, let's change to a better one if possible (function and flag name that is)

### CODE REVIEW

File: include/net/buf.h [hunk 1]
```diff
@@ -455,6 +455,16 @@ static inline void net_buf_simple_restore(struct net_buf_simple *buf,
...
 #define NET_BUF_EXTERNAL_DATA  BIT(1)
 
+/** Flag indicating that both the buffer and its associated data pointer,
+ * point to externally allocated memory and do not come from any pool.
+ * Therefore once ref goes down to zero, the pointed  buffer and data
+ * will not need to be deallocated. This never needs to be
+ * explicitly set or unset by the net_buf API user. Such net_buf is
+ * exclusively instantiated via net_buf_static_setup() function..
+ */
+#define NET_BUF_STATIC BIT(2)
```
Comment (jukkar) [inline include/net/buf.h]: There already exists NET_BUF_SIMPLE_DEFINE_STATIC(), can't we use that here instead?

Comment (tbursztyka) [reply include/net/buf.h]: It's a net_buf_simple, not a net_buf. So you won't be able to add/insert it into a list of other net_buf.

The idea is really to have a net_buf, play with it exactly as if it was allocated through net_buf_alloc or else, and getting it to work with others seamlessly

File: include/net/buf.h [hunk 2]
```diff
@@ -455,6 +455,16 @@ static inline void net_buf_simple_restore(struct net_buf_simple *buf,
...
+
+
```
Comment (jhedberg) [inline include/net/buf.h]: Redundant empty line

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks fine to me, except for the one minor whitespace (empty line) issue

**Comment by tbursztyka** [comment]
@jhedberg what about the name? is "static" a good one. Well it was used already for net_buf_simple actually. I'll add a macro NET_BUF_DEFINE_STATIC() as well, it can be useful for those who don't need to select a different data array at runtime.

**Comment by pfalcon** [comment]
> Instead of flag name NET_BUF_STATIC should it be instead NET_BUF_EXTERNAL?

Maybe. There's NET_BUF_EXTERNAL_DATA already, and then if both the struct and data it points to are "external" (to standard pool structures), makes sense to call the flag NET_BUF_EXTERNAL. The risk is confusion between "EXTERNAL_DATA" and "EXTERNAL" (a necessary/acceptable risk IMHO).

NET_BUF_STATIC would still work too, but indeed, there's confusion between a buffer being conceptually "static" and it and its data array being defined as static vars (that's slightly bigger confusion than the above IMHO).


Another way to deal with it would be to "decompose" settings to the most atomic properties, then define useful higher-level buffer types in terms of them. That's how lwIP does it for example. (I mention that for completeness, as a reference to "how others do that". I don't advocate doing that now, but I'm sure we'll need to define more buffer flags in the future (e.g. a flag for constant (ROM) vs volatile data). In this case, having atomic, orthogonal flags makes sense).

~~~
  PBUF_RAM = (PBUF_ALLOC_FLAG_DATA_CONTIGUOUS | PBUF_TYPE_FLAG_STRUCT_DATA_CONTIGUOUS | PBUF_TYPE_ALLOC_SRC_MASK_STD_HEAP),
  PBUF_ROM = PBUF_TYPE_ALLOC_SRC_MASK_STD_MEMP_PBUF,
  PBUF_REF = (PBUF_TYPE_FLAG_DATA_VOLATILE | PBUF_TYPE_ALLOC_SRC_MASK_STD_MEMP_PBUF),
  PBUF_POOL = (PBUF_ALLOC_FLAG_RX | PBUF_TYPE_FLAG_STRUCT_DATA_CONTIGUOUS | PBUF_TYPE_ALLOC_SRC_MASK_STD_MEMP_PBUF_POOL)
~~~

**Comment by tbursztyka** [comment]
@pfalcon confusion could be a risk, though a minor one. But why not NET_BUF_EXTERNAL_FULL instead of NET_BUF_EXTERNAL then? at least it's easy to grasp the difference with NET_BUF_EXTERNAL_DATA

Yep I recall the idea of having a flag for ROM based data. But that will need first a kernel api function for checking if pointer is in ROM or not. A worthy idea non the less.

**Comment by pfalcon** [comment]
> confusion could be a risk, though a minor one. But why not NET_BUF_EXTERNAL_FULL instead of NET_BUF_EXTERNAL then? at least it's easy to grasp the difference with NET_BUF_EXTERNAL_DATA

I don't have a specific preference here. NET_BUF_EXTERNAL_FULL sounds a bit "artificial", but indeed clearly distinguishable from NET_BUF_EXTERNAL_DATA.

**Comment by tbursztyka** [comment]
let's try NET_BUF_ALL_EXTERNAL

**Comment by jhedberg** [comment]
FWIW, I have nothing against calling it STATIC. If I understand right, this is just a buffer that's not associated with any pool, i.e. a standalone one. Calling it "external" sounds quite unintuitive to me. If I were to brainstorm sometihng else than STATIC, NET_BUF_STANDALONE or NET_BUF_NO_POOL come to mind.

**Comment by tbursztyka** [comment]
@jhedberg STANDALONE sounds nice, will do the change

File: include/net/buf.h [hunk 3]
```diff
@@ -455,6 +455,54 @@ static inline void net_buf_simple_restore(struct net_buf_simple *buf,
...
+/** @def NET_BUF_DEFINE
+ *  @brief Define a net_buf stack variable.
+ *
+ *  This is a helper macro which is used to define a net_buf object
+ *  on the stack.
+ *
+ *  @param _name Name of the net_buf object.
+ *  @param _size Maximum data storage for the buffer.
+ */
+#define NET_BUF_DEFINE(_name, _size)				\
```
Comment (jhedberg) [inline include/net/buf.h]: Are you really seeing the need for this? I'd imagine only the static variant to be useful, since if you need something only for the duration for the function's lifetime (that the variable is declared in) then net_buf_simple should be the way to go.

Comment (jhedberg) [reply include/net/buf.h]: Also, I'm a bit uncomfortable with how standard/normal the name of this looks like, considering that this is a bit of a corner case that will only have a few users. Meaning, it might be better to have "STANDALONE" in the API name, to make it clear that this is something "special" and not something that most users of net_buf will need.

**Comment by jhedberg** [comment]
One other question: since these buffers have no "deallocation". How are you planning to write safe code that will only try to reuse the buffer (manipulate the data, place it in a FIFO, etc) when no-one else uses it anymore? Since net_buf's have reference counting there might be arbitrary modules that the buffer passes through that grab a hold of the buffer with net_buf_ref, and assume they have valid access to it until they do net_buf_unref.

File: include/net/buf.h (hunk 3 above)
Comment (pfalcon) [reply include/net/buf.h]: @jhedberg 

> then net_buf_simple should be the way to go.

net_buf_simple is NOT used anywhere in the IP stack code. As it stands currently, net_buf_simple can as well be renamed to net_buf_adhoc_variant_for_bt_subsys ;-).

Comment (jhedberg) [reply include/net/buf.h]: I'm questioning here the introduction of an on-stack net_buf variable. I don't think there's so far been any use cases for that in the network subsystem. With such a variable you want to be sure that no-one has grabbed references to it by the time you return from your function (in whose scope the variable was declared in). net_buf_simple gives that, since there are no APIs (like net_buf_ref) that would give other functions the idea that they can "own" or hold onto the buffer.

As for the static net_buf variable, that's something I'm less concerned of, but I still fail to see how we can avoid buggy code if there's no way to catch the point when the buffer can be reused (and if the code has to wait for this, then I don't see a different with having a dedicated one-buffer net_buf_pool).

Btw, any net_buf using code does use net_buf_simple since that's how the data is internally stored in net_buf. This is particularly convenient when you have a net_buf and want to pass the data to some sub-function, but not give it the opportunity to grab a reference or mess with other net_buf state. In such a case you just pass `&buf->b`, i.e. a pointer to the net_buf_simple.

Comment (pfalcon) [reply include/net/buf.h]: > With such a variable you want to be sure that no-one has grabbed references to it by the time you return from your function

For reference, my understanding how this is handled is that refcounted buffers are counter-incremented, and of non-refcounted, a copy is made.

> I'm questioning here the introduction of an on-stack net_buf variable.

My understanding of this patch is that @tbursztyka wants to introduce a complete interface, to be available to any client whenever (if) needed (all prompted by his larger WIP refactor of buf handling in the IP code).

> Btw, any net_buf using code does use net_buf_simple since

Cool, and yet ident "net_buf_simple" is never referenced in subsys/net/ip/.


(All these responses are my trying to understand these (fairly complex, if not confusing) matters in grey-box manner. I definitely don't argue some point. (Except that all stuff should be understandable by other parties of course ;-) )).

Comment (jukkar) [reply include/net/buf.h]: > Cool, and yet ident "net_buf_simple" is never referenced in subsys/net/ip/.

There has been so far no need to use the net_buf_simple API in IP stack side of the code.

**Comment by tbursztyka** [comment]
Usage of such net_bug should be restricted to very particular usage yes.
For instance, in my case, I want to create L2 header right before sending the packet into such net_buf. In this particular case it's totally fine since on sending won't grab any reference anywhere there. 

But indeed, maybe that needs to be properly documented?

**Comment by jhedberg** [comment]
> For instance, in my case, I want to create L2 header right before sending the packet into such net_buf. In this particular case it's totally fine since on sending won't grab any reference anywhere there.

@tbursztyka could you elaborate on this a bit? I remember you saying the buffer would be placed in some FIFO (which is why net_buf_simple is no good), which means it'll be processed at some later moment, presumably by some other thread. How does the code that took this buffer and passed it onward know when it can be used again?

**Comment by pfalcon** [comment]
> I remember you saying the buffer would be placed in some FIFO

I'm sure @tbursztyka will have a chance to reply before he goes on vacation (which he said is today unless I missed it), but let me flex again my understanding of the matter (based on discussion with him): this functionality (standalone buffer) will be used when a packet is extracted from fifo by tx thread, which then processes it in synchronous manner. E.g. higher-level stack layers will prepare a complete IP packet and put it in fifo. TX thread will extract it, prepend L2 header using such a standalone/static net_buf, and make a call to driver to transmit a chain of buffers. When this call returns, the static buffer can be reused.

In other words, this standalone buffer is not expected to be an application-level construct, but rather an internal one.

**Review by pfalcon — APPROVED** [review]
Looks good to me at this stage. Thanks for submitting it separately and beforehand - this allows to understand the changes better, have a focused discussion (instead of discussing gazillion of things in 20-commits PR), and allows more developers be in loop and have better anticipation and picture of the upcoming changes.

**Comment by tbursztyka** [comment]
@jhedberg it's not going to be placed in FIFO, it's going to be appended/inserted into a net_buf list. And it is not going to be processed later (or then it would be a serious problem). 
Here is the idea, for instance on ethernet l2 (do not try to follow current code, it's going to change):

net_pkt being sent:
  - net_if: queue pkt
  - net_if_tx: use L2's send()  (TX thread at this point)
  - ethernet_send: create ethernet header in its locally allocated net_buf. Insert into net_pkt, send it to driver to send it.
  - driver: send the packet right away (no other thread, nothing, we are still in TX thread)
  - drivers returnse, either success or error: ethernet l2 removes the locally allocated net_buf from net_pkt
  - L2's send returns

**Comment by jhedberg** [comment]
@tbursztyka ok, thanks for the clarification. Then I would make at least a few things to avoid bugs with it:

 - Document this properly
 - Many API calls such as net_buf_ref, net_buf_unref, net_buf_put & net_buf_get fail if called for such a buffer.

**Comment by jhedberg** [comment]
Actually thinking more about this, what I suggested just opens a can of worms. E.g. until now net_buf_ref() has always succeeded and there's lots of code assuming it will. We'd need to update all those places if it can now suddenly return NULL for a buffer with the STANDALONE flag set.

Also, having the buffer in a linked list is a reference (in the sense that there's a pointer that references the buffer), and not incrementing the refcount for it is just a hack. What about the other buffers in the linked list which don't have the STANDALONE flag set - sounds like there might have been a bug there from the beginning if the linked list links don't use the refcounting mechanism.

So in summary, I'm starting to feel like the least hackish option would still be to have dedicated pool with a single buffer, and have an assert() if allocating from the pool fails using K_NO_WAIT, since that means that some code in the code path of the function that's supposed to consume the buffer hasn't actually fully consumed it.

Does @jukkar have some thoughts on this?

**Comment by jukkar** [comment]
> Make API calls such as net_buf_ref, net_buf_unref, net_buf_put & net_buf_get fail if called for such a buffer.

This is a bit problematic if we would need to check in the driver if the buf is "static" one or not before trying to unref if.

> So in summary, I'm starting to feel like the least hackish option would still be to have dedicated pool with a single buffer, and have an assert() if allocating from the pool fails using K_NO_WAIT, since that means that some code in the code path of the function that's supposed to consume the buffer hasn't actually fully consumed it.

Having a pool of just one buf for this purpose sounds like a good solution. This way this "special" buffer would look exactly like any other buffer, and we could control the life cycle of it very well.

**Comment by tbursztyka** [comment]
Just getting back on this.

Having a pool for a statically allocated net_buf is way too overkill. Idea of such net_buf is to use those very locally, and obviously knowing when a ref/unref is going to happen no avoid issues related to it.

I might drop that PR altogether, and instead up to - for instance - ethernet l2 to allocate statically a net_buf by itsels, put into the list, and removing it relevantly. (@jukka: there will not be any unref on driver side anymore btw). It's not going to be super clean but at least will avoid any ambiguity .

**Comment by tbursztyka** [comment]
@jhedberg, @jukkar  fyi: https://github.com/tbursztyka/zephyr/commit/e6d99789d493d3ddbd2eaa4008442bf8b3a6fa7b Well, it works pretty well that way, no need to temper with net_buf code base then.
